### PR TITLE
Created models for JPA

### PR DESCRIPTION
**Pull request by tomrsae**

### CODE REVIEW

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/controller/LoginController.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+
+@RestController
+@RequestMapping(value = "/login")
+@EnableAutoConfiguration
+@CrossOrigin(origins = "WRITE HOST HERE") //TODO:
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/controller/LoginController.java]: Should read from an env var

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Assignment.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+public class Assignment {
+    @Id
+    private int id;
+
+    private String name;
+
+    @ManyToOne
+    private Subject subject;
+
+    public Assignment(int id, String name, Subject subject) {
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Assignment.java]: add not null

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/LoginRequest.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+public class LoginRequest {
+
+    private final String email;
+    private final String password;
+
+    @JsonCreator
+    public LoginRequest(String email, String password) {
+        this.email = email;
+        this.password = password;
+    }
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/LoginRequest.java]: add null-checking

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/LoginResponse.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package com.github.stackscrubs.stuq.backend.model;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/LoginResponse.java:9]: Can't we use optional instead?

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Student.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+	public Student( int id,
+                    String firstName,
+                    String lastName,
+                    String email,
+                    String phone,
+                    String password_hash)
+    {
+		super(id, firstName, lastName, email, phone, password_hash);
+	}
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Student.java]: not null here too? :)

Comment (ngorke) [reply backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Student.java]: not sure if you need it but the JPA not null thing :P

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Subject.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import javax.persistence.JoinTable;
+import javax.persistence.ManyToMany;
+
+@Entity
+public class Subject {
+    @Id
+    private String code;
+
+    @Id
+    private Term term;
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Subject.java]: non nullable?

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Submission.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    private Student student;
+
+    @Id
+    private Assignment assignment;
+
+    private boolean is_approved;
+
+    private String feedback;
+
+    public Submission(Student student, Assignment assignment, boolean is_approved, String feedback) {
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Submission.java]: use camelCase

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Submission.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        this.student = student;
+        this.assignment = assignment;
+        this.is_approved = is_approved;
+        this.feedback = feedback;
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Submission.java]: set correct nullability

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Term.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+import javax.persistence.Entity;
+import javax.persistence.Id;
+
+@Entity
+public class Term {
+    @Id
+    private int year;
+    
+    @Id
+    private String period;
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/Term.java]: nullability! don't we need getters as well?

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/User.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+                String phone,
+                String password_hash)
+    {
+        this.id = id;
+        this.firstName = firstName;
+        this.lastName = lastName;
+        this.email = email;
+        this.phone = phone;
+        this.password_hash = password_hash;
+    }
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/User.java]: non-nullable! 😄

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/service/LoginService.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package com.github.stackscrubs.stuq.backend.service;
+
+import com.github.stackscrubs.stuq.backend.model.LoginRequest;
+import com.github.stackscrubs.stuq.backend.model.LoginResponse;
+
+public class LoginService {
+
+    public LoginResponse login(LoginRequest login) {
+        // TODO!
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/service/LoginService.java]: todo 👍

Comment (ngorke) [reply backend/src/main/java/com/github/stackscrubs/stuq/backend/service/LoginService.java]: an exception would be preferable, as then we know for sure we forgot to implement it

**Review by ngorke — CHANGES_REQUESTED** [review]
Puuurty good job! 🦖  I commented on some small things 💯

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/controller/LoginController.java (hunk 1 above)
Comment (tomrsae) [reply backend/src/main/java/com/github/stackscrubs/stuq/backend/controller/LoginController.java]: Mapping too, or only origins?

File: backend/.env [hunk 12]
```diff
@@ -0,0 +1 @@
+HOST=
```
Comment (ngorke) [inline backend/.env]: can remove file

**Review by tomrsae — COMMENTED** [review]
Make use of `this` consistent accross all files.

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/jpa/Student.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+	public Student( @NonNull int id,
+                  @NonNull String firstName,
+                  @NonNull String lastName,
+                  String email,
+                  String phone,
+                  @NonNull String password_hash)
+  {
+		super(id, firstName, lastName, email, phone, password_hash);
+	}
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/jpa/Student.java]: these brackets are looking kind of unaligned

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/jpa/SubjectId.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        return term;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + ((code == null) ? 0 : code.hashCode());
+        result = prime * result + ((term == null) ? 0 : term.hashCode());
+        return result;
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/jpa/SubjectId.java]: use Objects.hash

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/jpa/SubmissionId.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        return assignment;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + ((assignment == null) ? 0 : assignment.hashCode());
+        result = prime * result + ((student == null) ? 0 : student.hashCode());
+        return result;
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/jpa/SubmissionId.java]: Object.hash :)

File: backend/src/main/java/com/github/stackscrubs/stuq/backend/model/jpa/TermId.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        return period;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + ((period == null) ? 0 : period.hashCode());
+        result = prime * result + year;
+        return result;
```
Comment (ngorke) [inline backend/src/main/java/com/github/stackscrubs/stuq/backend/model/jpa/TermId.java]: Object.hash

**Review by ngorke — CHANGES_REQUESTED** [review]
R🧅

**Review by ngorke — APPROVED** [review]
noice

**Pull request merged**